Controller/StaticPageAdminController.php

```diff
@@ -46,69 +39,46 @@ public function treeAction()
-
-        $iterator = new \RecursiveIteratorIterator(
-            new \RecursiveArrayIterator($newTree),
-            \RecursiveIteratorIterator::LEAVES_ONLY
-        );
-
-        $rootStaticPage = $staticPageRepo->findRootStaticPage();
-        $lastStaticPage = $rootStaticPage;
-        $parents        = [
-            0 => $rootStaticPage
-        ];
-        $position       = 1;
-        $depth          = 1;
-        $lastDepth      = $depth;
-
-        foreach ($iterator as $staticPageId) {
-            $depth = round(($iterator->getDepth() - 1)/2);
-
-            if ($depth > $lastDepth) {
-                $parents[$depth] = $lastStaticPage;
-            }
-
-            $staticPage = $staticPageRepo->find($staticPageId);
+        /**
+         * @var StaticPageManager $staticPageManager
+         */
+        $staticPageManager = $this->get('grossum_static_page.static_page.manager');
 
-            $staticPage->setParent($parents[$depth]);
-            $staticPage->setPosition($position);
+        $root = $staticPageManager->getRepository()->findRootStaticPage();
 
-            $lastStaticPage = $staticPage;
-            $lastDepth      = $depth;
-
-            ++$position;
-        }
+        foreach($newTree as $treeData) {
+            if (isset($treeData['item_id']) && $treeData['item_id'] === StaticPage::ROOT) {
+                continue;
+            }
 
-        $connection = $this->getDoctrine()->getConnection();
-        /* @var $connection Connection */
+            if (!isset($treeData['parent_id']) || !isset($treeData['id'])) {
+                continue;
+            }
 
-        // GedmoTree-extension make changes in database without transactions, but we want to have not broken tree,
-        // that's why we manage transaction manually
-        try {
-            $connection->beginTransaction();
+            /**
+             * @var StaticPage $staticPage
+             */
+            $staticPage = $staticPageManager->getRepository()->find($treeData['id']);
+            $parentId = ($treeData['parent_id'] === StaticPage::ROOT) ? $root->getId() : $treeData['parent_id'];
+            $parentStaticPage = $staticPageManager->getRepository()->find($parentId);
 
-            $this->getDoctrine()->getManager()->flush();
-            $staticPageRepo->reorder($rootStaticPage, 'position');
-            $this->getDoctrine()->getManager()->flush();
+            $staticPage->setParent($parentStaticPage)
+                ->setLft($treeData['left'])
+                ->setRgt($treeData['right']);
+        }
 
-            $connection->commit();
-        } catch (\Exception $e) {
-            $connection->rollBack();
+        $verified = $staticPageManager->getRepository()->verify();
 
-            throw $e;
+        if ($verified !== true) {
+            $this->getDoctrine()->getManager()->clear();
+            return new JsonResponse(['result' => false]);
         }
+        $staticPageManager->flush();
```

Comment:
I think this:
`$verified = $staticPageManager->getRepository()->verify();`
and this:
`$staticPageManager->flush();`
must be in one transaction like this:
```
       try {
            $connection->beginTransaction();

            $staticPageManager->getRepository()->verify();
            --//--//--
            $staticPageManager->flush();
         
            $connection->commit();
        } catch (\Exception $e) {
            $connection->rollBack();

            throw $e;
        }
```
because `->verify()` don't use transaction